sdk/rust/oak/src/http/mod.rs

```diff
@@ -21,24 +21,35 @@ use crate::{
     proto::oak::invocation::HttpInvocationSender,
     OakError, OakStatus,
 };
+use http::{Request, Response};
 use log::warn;
-pub use oak_services::proto::oak::encap::{HttpRequest, HttpResponse};
+pub use oak_services::proto::oak::encap::{HeaderMap, HttpRequest, HttpResponse};
+use std::convert::TryInto;
 
 pub type Invocation = crate::proto::oak::invocation::HttpInvocation;
 
+/// This implementation provides an interface for sending requests and receiving responses, using
+/// the idiomatic http types. Internally, these types are converted into protobuf encoded requests
+/// and responses that can be used for communication with the Oak nodes.
 impl Invocation {
-    pub fn receive(&self) -> std::result::Result<HttpRequest, crate::OakError> {
-        self.receiver
+    pub fn receive(&self) -> std::result::Result<Request<Vec<u8>>, crate::OakError> {
+        let request = self
+            .receiver
             .as_ref()
             .ok_or(OakError::OakStatus(OakStatus::ErrBadHandle))?
-            .receive()
+            .receive();
+
+        request
+            .map(|req| req.try_into().map_err(crate::OakError::OakStatus))
+            .flatten()
```

Comment:
FYI, if `filter_map` works here, it is usually more idiomatic.

Reply:
I think only `Iter` has `flat_map`, I cannot find `flat_map` for `Result`.

Reply:
`flat_map` for Result and Option is called `and_then`

Reply:
In case you are interested, the reason that all those functions "feel" related is that they are actually are instances of Monads, and `flat_map` / `and_then` is the "bind" operator. See http://learnyouahaskell.com/a-fistful-of-monads .

Reply:
Oh! That helps. Thanks :)